WIP: Build via Containerfile and derive from fedora-bootc

jlebon opened the pull request:
This uses https://github.com/coreos/rpm-ostree/pull/5274 to build FCOS as a derivation of fedora-bootc.

jlebon commented:
This is still in the early phase. It doesn't work yet.

Also note we're using tier-1 for now since there's no tier-x image available yet (https://gitlab.com/fedora/bootc/base-images/-/issues/25).

jlebon commented:
OK cool, got a successful build with this and https://github.com/coreos/rpm-ostree/pull/5274 which I can even rebase to and boot successfully. We're not rebuilding the initramfs yet nor adding overlays. Those shouldn't be hard though.

dustymabe commented:
Havne't had a chance to test this out yet. Some comments in the mean time.

jlebon commented on build-rootfs:
```diff
@@ -0,0 +1,239 @@
+        path = os.path.join(CONTEXTDIR, 'overlay.d', layer[len('overlay/'):])
+        shutil.copytree(path, rootfs, dirs_exist_ok=True)
+
+    # inject live/ bits
+    target_path = os.path.join(rootfs, 'usr/share/coreos-assembler/live')
+    shutil.copytree(os.path.join(CONTEXTDIR, "live"), target_path)
```
Yeah, true. It's trivial too to just inject it so will just do that for now to avoid the cosa churn.

jlebon commented:
Updated for comments!

This now requires https://gitlab.com/fedora/bootc/base-images/-/merge_requests/195.

Split prep patches in https://github.com/coreos/fedora-coreos-config/pull/3534.

Dropped the `WIP` prefix to signal this is nearing completion.

cgwalters commented:
Thanks again for all of the work on this! It's a giant milestone towards the original vision I had here...been a long, long journey and many parts have been much harder than I thought. But I do believe (hope) that this will pay itself back in a lot of ways. 

I mean, one thing *for sure* that is so much dramatically nicer is that use case of "hey I just want to drop in this one override for one package" is now just `FROM <base> RUN dnf -y upgrade foo` whereas before...so much bespoke tooling.

If there's anything more I (or the team) can do to help let me know!

jlebon commented:
OK, updated and dropped out of draft!

The new GHA will fail until the next Konflux run that picks up https://gitlab.com/fedora/bootc/base-images/-/merge_requests/195.

dustymabe commented on build-rootfs:
I agree with you that in the context of this file it doesn't seem necessary to name it more specifically, but the cache type mounts are all stored under `/var/tmp` and having it be named something more specific can be useful there.

dustymabe commented on Containerfile:
```diff
@@ -0,0 +1,36 @@
+RUN --mount=type=bind,target=/run/src /run/src/build-rootfs "${MANIFEST}" "${VERSION}" /target-rootfs
+
+FROM scratch
+ARG VERSION
```
dumb question.. why do we have `ARG VERSION` 3 times in this file? I guess the context gets reset and naming it again is useful?

jlebon replied:
There are scoping rules with `ARG` that required me to redefine them when I was hacking on this, but yeah looking at the final state I think I should be able to get rid of some of them.

dustymabe commented on Containerfile:
```diff
@@ -0,0 +1,36 @@
+ARG VERSION="42"
```
Considering this file will get synced to other branches of this repo should we not be hardcoding `42` here and also in the `FROM` command below this but rather pull that from the manifest somehow.

jlebon replied:
The immediate problem with this is that it requires `cpp` which isn't in FCOS and so not in our builders either. While `cpp` is in cosa, it would be quite unfortunate more generally to require users to have `cpp`.

So moved away from that and just switched to requiring us to concatenate the pieces ourselves and committing that. I'm not really happy with this, but so far it looks like the least terrible option.

cgwalters replied:
Up to you all of course but my lean on things like this is to just have a plain `podman build` do something useful (it's ok if the version numbers aren't what's expected) but I would expect to have a `Justfile` or something equivalent that helps me to trivially locally what a more "production/orchestrated" pipeline would do.

jlebon replied:
Definitely possible, yeah. I think it might eventually happen, though I want to try to avoid adding a `Makefile`/`Justfile` into this repo, and always keeping `podman build` the canonical entrypoint since it works as a "choke point" through which we must ensure all enhancements are implemented.

Dynamic label information (like versions) is one of those that you kinda have to drive from the outside (though https://github.com/containers/buildah/pull/5975 in theory would allow it), but there really shouldn't be that many such cases.

dustymabe commented on build-rootfs:
```diff
@@ -0,0 +1,229 @@
+#!/usr/bin/python3
+
```
optional: might be nice to have a description of what this file is for (what it does) here.

dustymabe reviewed (approved):
LGTM

jlebon commented on Containerfile.base:
```diff
@@ -0,0 +1,31 @@
+ARG VERSION
+ARG MANIFEST
+
+# useful if you're hacking on rpm-ostree/bootc-base-imagectl
+# COPY rpm-ostree /usr/bin/
+# COPY bootc-base-imagectl /usr/libexec/
```
I like that, though iteration speed-wise it seems less efficient.

dustymabe commented:
For some reason in my local desktop I can't get this to build. I'm seeing the following when I `podman build --security-opt=label=disable --cap-add=all --device /dev/fuse -t localhost/fedora-coreos .` locally:

```
Creating empty machine-id
error: Executing compose install: Subprocess failed: ExitStatus(unix_wait_status(256))
p/dracut/dracut.29iPSx/initramfs /usr/lib/dracut/modules.d/35coreos-live/coreos-livepxe-persist-osmet.service /usr/lib/systemd/system/coreos-livepxe-persist-osmet.service
cp: setting attributes for '/tmp/dracut/dracut.29iPSx/initramfs/lib/modules/6.14.9-300.fc42.x86_64//kernel/drivers/block/loop.ko.xz': Operation not supported
dracut-install: ERROR: 'cp --reflink=auto --sparse=auto --preserve=mode,xattr,timestamps,ownership -fL /lib/modules/6.14.9-300.fc42.x86_64//kernel/drivers/block/loop.ko.xz /tmp/dracut/dracut.29iPSx/initramfs/lib/modules/6.14.9-300.fc42.x86_64//kernel/drivers/block/loop.ko.xz' failed with 1
dracut-install: dracut_install 'loop' ERROR
dracut-install: ERROR: installing 'loop'
dracut[E]: FAILED:  /usr/lib/dracut/dracut-install -D /tmp/dracut/dracut.29iPSx/initramfs --kerneldir /lib/modules/6.14.9-300.fc42.x86_64/ -m loop
dracut[F]: installkernel failed in module coreos-live
```

that being said, @jlebon isn't able to reproduce. I'm happy to merge this in and debug my setup later.

jbtrystram commented:
FYI i was able to build that locally but i had to override the from with : 
`quay.io/bootc-devel/fedora-bootc-42-standard:git-f7d48642c044ccfa89f9c0b9c18871f0f1ed1e8d`

jlebon commented:
> FYI i was able to build that locally but i had to override the from with : `quay.io/bootc-devel/fedora-bootc-42-standard:git-f7d48642c044ccfa89f9c0b9c18871f0f1ed1e8d`

@jbtrystram Hmm weird, the latest `quay.io/bootc-devel/fedora-bootc-42-standard` should've worked. What error were you seeing?

jbtrystram commented:
bootc-base-imagebuilder doesn't know about `--add-dir` on latest :

```
 [~] podman run --rm -it --pull=always quay.io/bootc-devel/fedora-bootc-42-standard:latest grep add-dir /usr/libexec/bootc-base-imagectl                            
Trying to pull quay.io/bootc-devel/fedora-bootc-42-standard:latest...
Getting image source signatures
Copying blob c22a22f9973e skipped: already exists  
Copying config f2fd0d907f done   | 
Writing manifest to image destination
 [~]                                                                                                                                                                
 [~] podman run --rm -it --pull=always quay.io/bootc-devel/fedora-bootc-42-standard:git-f7d48642c044ccfa89f9c0b9c18871f0f1ed1e8d grep add-dir /usr/libexec/bootc-base-imagectl
Trying to pull quay.io/bootc-devel/fedora-bootc-42-standard:git-f7d48642c044ccfa89f9c0b9c18871f0f1ed1e8d...
Getting image source signatures
Copying blob a975329ad19c skipped: already exists  
Copying config 3da1278f65 done   | 
Writing manifest to image destination
    build_rootfs.add_argument("--add-dir", help='Copy dir contents into the target', action='append', default=[], metavar='DIR')
```

which is surprising given that `latest` image have been updated recently, but i haven't looked at the konflux definitions, maybe latest is only updated on a git tag or a manual release ?

jbtrystram commented:
ah, looking at the labels on `latest` I see : 
```json
"Labels": {
        "architecture": "x86_64",
        "bootc.diskimage-builder": "quay.io/centos-bootc/bootc-image-builder",
        "build-date": "2025-06-04T14:01:42",
        "containers.bootc": "1",
        "io.buildah.version": "1.39.0-dev",
        "org.opencontainers.image.version": "43",
        "vcs-ref": "08254c0ba435bed9f8871a0c6b0ed806599ed41c",
        "vcs-type": "git"
}
```

haven't been rebuilt since june 04th

cgwalters commented:
```
$ podman run --rm -ti quay.io/fedora/fedora-bootc:42 /usr/libexec/bootc-base-imagectl --help                                                        2 06/11/2025 09:29:50 AM
usage: bootc-base-imagectl [-h] [--args-file FILE] {build-rootfs,rechunk,list} ...
```

(And thank you for that!!!)

mmartinv commented:
> ah, looking at the labels on `latest` I see :
> 
> ```json
> "Labels": {
>         "architecture": "x86_64",
>         "bootc.diskimage-builder": "quay.io/centos-bootc/bootc-image-builder",
>         "build-date": "2025-06-04T14:01:42",
>         "containers.bootc": "1",
>         "io.buildah.version": "1.39.0-dev",
>         "org.opencontainers.image.version": "43",
>         "vcs-ref": "08254c0ba435bed9f8871a0c6b0ed806599ed41c",
>         "vcs-type": "git"
> }
> ```
> 
> haven't been rebuilt since june 04th

This is basically my fault: Yesterday I was trying to clean up the failed releases and I accidentally triggered some of them which, for what it seems, were quite old. This should have been fixed with the latest builds from yesterday's compose
